In [2]:
import pandas as pd
from pathlib import Path
import re
from rapidfuzz import fuzz, process
from geopy.geocoders import Nominatim
import us

BASE_DIR = Path("..")
ORIGINAL_DATA_DIR = BASE_DIR / "original_data"
CLEAN_DATA_DIR = BASE_DIR / "clean_data"
FILE_PATH = ORIGINAL_DATA_DIR / "nsf_terminations_airtable.csv"
TARGET_PATH = CLEAN_DATA_DIR / "nsf_terminations_airtable.csv"
def describe_dataset(df):
    summary = []
    for col in df.columns:
        col_data = df[col]
        summary.append({
            'Column': col,
            'Type': col_data.dtype,
            'Nulls': col_data.isnull().sum(),
            'Null %': round(col_data.isnull().mean() * 100, 2),
            'Unique': col_data.nunique(dropna=True),
            'Min': col_data.min() if pd.api.types.is_numeric_dtype(col_data) else None,
            'Max': col_data.max() if pd.api.types.is_numeric_dtype(col_data) else None,
            'Mean': round(col_data.mean(), 3) if pd.api.types.is_numeric_dtype(col_data) else None,
            'Std': round(col_data.std(), 3) if pd.api.types.is_numeric_dtype(col_data) else None,
            'Example values': col_data.dropna().unique()[:3].tolist()  # first 3 unique examples
        })
    
    summary_df = pd.DataFrame(summary)
    return summary_df



In [3]:
data = pd.read_csv(FILE_PATH)

display(data.head(5))

grant_id                 status  terminated  suspended termination_date  \
0   2201796           ❌ Terminated        True      False       2025-04-18   
1   2315095           ❌ Terminated        True      False       2025-04-18   
2   2236163           ❌ Terminated        True      False       2025-04-18   
3   2243822  🔄 Possibly Reinstated        True      False       2025-04-18   
4   2216648           ❌ Terminated        True      False       2025-04-18   

   reinstated reinstatement_date     reinstatement_indicator  \
0       False                NaN                         NaN   
1       False                NaN                         NaN   
2       False                NaN                         NaN   
3        True         2025-06-30  Thakur v. Trump (Jun 2025)   
4       False                NaN                         NaN   

                                             nsf_url  \
0  https://www.nsf.gov/awardsearch/showAward?AWD_...   
1  https://www.nsf.gov/awardsearch/showAward?AWD_...   
2  https://www.nsf.gov/awardsearch/showAward?AWD_...   
3  https://www.nsf.gov/awardsearch/showAward?AWD_...   
4  https://www.nsf.gov/awardsearch/showAward?AWD_...   

                                     usaspending_url  ...  \
0  https://www.usaspending.gov/award/ASST_NON_220...  ...   
1  https://www.usaspending.gov/award/ASST_NON_231...  ...   
2  https://www.usaspending.gov/award/ASST_NON_223...  ...   
3  https://www.usaspending.gov/award/ASST_NON_224...  ...   
4  https://www.usaspending.gov/award/ASST_NON_221...  ...   

  usaspending_obligated usaspending_outlaid estimated_budget  \
0             1083218.0           901981.05          1083218   
1                   NaN                 NaN          3641116   
2              609429.0           348712.21           609429   
3                   NaN                 NaN          3000000   
4             2719700.0          1547300.07          2719700   

  estimated_outlays estimated_remaining  \
0      9.019811e+05        1.812369e+05   
1      3.641116e+06        0.000000e+00   
2      3.487122e+05        2.607168e+05   
3      1.079409e+06        1.920591e+06   
4      1.547300e+06        1.172400e+06   

                                            division          directorate  \
0  Research on Learning in Formal and Informal Se...       STEM Education   
1                                                NaN                  NaN   
2                            Undergraduate Education       STEM Education   
3                                                NaN                  NaN   
4                          Biological Infrastructure  Biological Sciences   

   div  dir                               record_sha1  
0  DRL  EDU  5866299f5d416a23e01823f57f857aad2950cd4d  
1  NaN  NaN  136bbe81751eb4f4ae26f6dceb68ee1339877daa  
2  DUE  EDU  075e2156a2996eba5595f0095b20da6da75c06e7  
3  NaN  NaN  f9a85b61f807953ecdb2a77c92362c042dcfb494  
4  DBI  BIO  f35dec9f84e3160f0ff2eee79200f25ff5471f67  

[5 rows x 35 columns]

In [4]:
describe_dataset(data)

Column     Type  Nulls  Null %  Unique             Min  \
0                  grant_id    int64      0    0.00    1970         1231319   
1                    status   object      0    0.00       2            None   
2                terminated     bool      0    0.00       1            True   
3                 suspended     bool      0    0.00       1           False   
4          termination_date   object      0    0.00      12            None   
5                reinstated     bool      0    0.00       2           False   
6        reinstatement_date   object   1550   78.68       4            None   
7   reinstatement_indicator   object   1554   78.88       2            None   
8                   nsf_url   object      0    0.00    1970            None   
9           usaspending_url   object      0    0.00    1970            None   
10            project_title   object      0    0.00    1658            None   
11                 abstract   object      1    0.05    1670            None   
12                 org_name   object      0    0.00     507            None   
13                org_state   object      0    0.00      52            None   
14                 org_city   object      1    0.05     329            None   
15               award_type   object      0    0.00       5            None   
16           usa_start_date   object    711   36.09     132            None   
17             usa_end_date   object    711   36.09      53            None   
18           nsf_start_date   object      0    0.00     145            None   
19             nsf_end_date   object      0    0.00      59            None   
20         nsf_program_name   object      1    0.05     293            None   
21      nsf_primary_program   object      0    0.00     586            None   
22           usa_nsf_office   object    711   36.09      34            None   
23         nsf_total_budget    int64      0    0.00    1642            6774   
24            nsf_obligated    int64      0    0.00    1727            6774   
25    usaspending_obligated  float64    711   36.09    1112             0.0   
26      usaspending_outlaid  float64    711   36.09    1170       -10217.22   
27         estimated_budget    int64      0    0.00    1681            6774   
28        estimated_outlays  float64      0    0.00    1874       -10217.22   
29      estimated_remaining  float64      0    0.00    1854 -1421008.105263   
30                 division   object    711   36.09      34            None   
31              directorate   object    711   36.09       9            None   
32                      div   object    711   36.09      34            None   
33                      dir   object    711   36.09       9            None   
34              record_sha1   object      0    0.00    1970            None   

            Max         Mean          Std  \
0       2531008  2243661.047   148987.307   
1          None          NaN          NaN   
2          True        1.000        0.000   
3         False        0.000        0.000   
4          None          NaN          NaN   
5          True        0.213        0.410   
6          None          NaN          NaN   
7          None          NaN          NaN   
8          None          NaN          NaN   
9          None          NaN          NaN   
10         None          NaN          NaN   
11         None          NaN          NaN   
12         None          NaN          NaN   
13         None          NaN          NaN   
14         None          NaN          NaN   
15         None          NaN          NaN   
16         None          NaN          NaN   
17         None          NaN          NaN   
18         None          NaN          NaN   
19         None          NaN          NaN   
20         None          NaN          NaN   
21         None          NaN          NaN   
22         None          NaN          NaN   
23     25000000   871612.779  1305502.115   
24     44434392   778344.299  1733191.316   
25   4

NULLS VALUES

Our hypothesis is that the 711 columns that have null values in usa_* are all the same for all variables that have 711 null. Let's check this.

In [5]:
usa_cols = [
    "usa_start_date", "usa_end_date", "usa_nsf_office",
    "usaspending_obligated", "usaspending_outlaid",
    "division", "directorate", "div", "dir"
]

mask = data[usa_cols].isna().all(axis=1)

print("Rows with all USA-related columns null:", mask.sum())

Rows with all USA-related columns null: 711


This code confirms this. This means that there are 711 entries that are on the nsf_* and not on the usa_*. This is not a problem, this means that we should use the 
"nsf_total_budget" variable and ignore the ones that have null for the 711 entries. The other null values make sense since they are reinstated dates of rows that have not been reinstated.
There is one more NULL value to deal with, there is one row with null on the abstract. Let's check what this is.

In [6]:
row_null_abstract = data[data["abstract"].isna()]

This seems like a normal row, for the moment wi will set the abstract to " " and keep with it.

In [7]:
data.loc[data["abstract"].isna(), "abstract"] = " "

We will first get the only columns that interest  us, we have noticed that terminated and suspended are always True so we can eliminate those columns also

In [8]:
selected_cols = [
    "grant_id",
    "status",
    "reinstated",
    "project_title",  
    "abstract",
    "org_name",
    "org_state",
    "nsf_total_budget"
]

data = data[[col for col in selected_cols if col in data.columns]]

We check for duplicades, which we realise that there are none.

In [9]:
data[data.duplicated(subset=["grant_id"], keep=False)]

Empty DataFrame
Columns: [grant_id, status, reinstated, project_title, abstract, org_name, org_state, nsf_total_budget]
Index: []

In [10]:
convert_types = {
    "grant_id": "Int64",                    
    "status": "category",                    
    "reinstated": "boolean",              
    "project_title": "string",                
    "abstract": "string",                    
    "org_name": "string",                     
    "org_state": "category",                  
    "usaspending_obligated": "float64"        
}


for col, dtype in convert_types.items():
    if col in data.columns:
        if dtype == "category":
            data[col] = data[col].astype("category")

        elif dtype == "boolean":
            data[col] = data[col].astype("boolean")

        elif dtype in ["Int64", "float64"]:
            data[col] = pd.to_numeric(data[col]).astype(dtype)

        elif dtype == "string":
            # Clean and normalize text fields
            if col in ["project_title", "abstract"]:
                def clean_to_string(text):
                    if pd.isna(text):
                        return ""
                    text = text.lower()
                    text = re.sub(r"[^a-z\s]", " ", text)
                    text = re.sub(r"\s+", " ", text)
                    return text.strip()

                data[col] = data[col].astype("string").apply(clean_to_string)

            else:
                data[col] = data[col].astype("string").str.strip()

In [11]:
if "status" in data.columns:
    data["status"] = (
        data["status"]
        .astype("string")
        .str.replace("❌", "", regex=False)
        .str.replace("🔄", "", regex=False)
        .str.strip()
        .replace({
            "Possibly Reinstated": "Reinstated",
        })
    )

In [12]:
describe_dataset(data)

Column            Type  Nulls  Null %  Unique      Min       Max  \
0          grant_id           Int64      0     0.0    1970  1231319   2531008   
1            status  string[python]      0     0.0       2     None      None   
2        reinstated         boolean      0     0.0       2    False      True   
3     project_title          object      0     0.0    1649     None      None   
4          abstract          object      0     0.0    1665     None      None   
5          org_name  string[python]      0     0.0     507     None      None   
6         org_state        category      0     0.0      52     None      None   
7  nsf_total_budget           int64      0     0.0    1642     6774  25000000   

          Mean          Std                                     Example values  
0  2243661.047   148987.307                        [2201796, 2315095, 2236163]  
1          NaN          NaN                           [Terminated, Reinstated]  
2        0.213        0.410                                      [False, True]  
3          NaN          NaN  [collaborative research investigating gender d...  
4          NaN          NaN  [despite evidence that gender differences in m...  
5          NaN          NaN  [Carnegie-Mellon University, University of Tex...  
6          NaN          NaN                                       [PA, TX, NM]  
7   871612.779  1305502.115                         [1051218, 3641116, 609429]

Let's change the USA state code to complete name so it mathces with us_state_population.csv. Note that
this check is also checking wether the sates are real and there are no invalid state names of inexistent states because the us_state_abbrev dictionary is completely correct.

In [13]:

us_state_abbrev = {
    'AL': 'Alabama', 'AK': 'Alaska', 'AZ': 'Arizona', 'AR': 'Arkansas',
    'CA': 'California', 'CO': 'Colorado', 'CT': 'Connecticut', 'DE': 'Delaware',
    'FL': 'Florida', 'GA': 'Georgia', 'HI': 'Hawaii', 'ID': 'Idaho',
    'IL': 'Illinois', 'IN': 'Indiana', 'IA': 'Iowa', 'KS': 'Kansas',
    'KY': 'Kentucky', 'LA': 'Louisiana', 'ME': 'Maine', 'MD': 'Maryland',
    'MA': 'Massachusetts', 'MI': 'Michigan', 'MN': 'Minnesota', 'MS': 'Mississippi',
    'MO': 'Missouri', 'MT': 'Montana', 'NE': 'Nebraska', 'NV': 'Nevada',
    'NH': 'New Hampshire', 'NJ': 'New Jersey', 'NM': 'New Mexico', 'NY': 'New York',
    'NC': 'North Carolina', 'ND': 'North Dakota', 'OH': 'Ohio', 'OK': 'Oklahoma',
    'OR': 'Oregon', 'PA': 'Pennsylvania', 'RI': 'Rhode Island', 'SC': 'South Carolina',
    'SD': 'South Dakota', 'TN': 'Tennessee', 'TX': 'Texas', 'UT': 'Utah',
    'VT': 'Vermont', 'VA': 'Virginia', 'WA': 'Washington', 'WV': 'West Virginia',
    'WI': 'Wisconsin', 'WY': 'Wyoming', 'DC': 'District of Columbia',
    'PR': 'Puerto Rico', 'GU': 'Guam', 'VI': 'Virgin Islands', 'AS': 'American Samoa'
}

data["org_state_full"] = data["org_state"].map(us_state_abbrev)

print(f"Unmatched rows: {len(data[data['org_state_full'].isna()])}")

print(len(us_state_abbrev.values()))

Unmatched rows: 0
55


Let's now check if there are any inconsistencies in the ORG name column, for this we will use rapidfuzz library with compares the
similarity of two names, we will display the results with 90 threshold, but we analysed it with and 80 one. The ones between 80 and 90 were not good matches.

In [14]:

orgs = sorted(data["org_name"].dropna().unique())
similar_pairs = []

for i, name in enumerate(orgs):
    matches = process.extract(name, orgs[i+1:], scorer=fuzz.token_sort_ratio, limit=5)
    for match, score, _ in matches:
        if score >= 90:  
            similar_pairs.append((name, match, score))

similar_df = pd.DataFrame(similar_pairs, columns=["org_name_1", "org_name_2", "similarity"])
display(similar_df.sort_values("similarity", ascending=False).head(30))


org_name_1                            org_name_2  \
4              Northeastern University               Northwestern University   
0               Boise State University                Bowie State University   
5       University of Colorado Boulder     University of Colorado at Boulder   
6             University of Washington                 Washington University   
2                     Clark University                   Clarkson University   
3                     Miami University                   University of Miami   
1  CUNY Kingsborough Community College  CUNY Queensborough Community College   

   similarity  
4   95.652174  
0   95.454545  
5   95.238095  
6   93.333333  
2   91.428571  
3   91.428571  
1   90.140845

The only two that should bother us are the "University of Colorado Boulder ---- University of Colorado at Boulder" and the "Miami University ---- University of Miami".
After doing some research online, the Miami ones correspond to the one in Ohio and the one in Florida. However the Boulder one seems to be the same university.

In [15]:
replacements = {
    "University of Colorado at Boulder": "University of Colorado Boulder"
}

data["org_name"] = data["org_name"].replace(replacements)

Let's now verify that the universities are in the states.

In [16]:
#DESCOMENTAR PER L'ENTREGA FINAL TRIGA MOLT I NO FA RES, NOMÉS DONA INFORMACIÓ
"""


geolocator = Nominatim(user_agent="nsf_org_checker")

def get_state_from_geopy(org_name):
    try:
        location = geolocator.geocode(org_name, addressdetails=True, timeout=10)
        if location and "address" in location.raw:
            address = location.raw["address"]
            state = address.get("state") or address.get("state_code")
            return state
    except Exception:
        pass
    return None

sample_df = data.copy() 

sample_df["real_state"] = sample_df["org_name"].apply(get_state_from_geopy)

def normalize_state_name(state):
    if not state:
        return None
    state_lookup = us.states.lookup(state)
    return state_lookup.name if state_lookup else state

sample_df["org_state_norm"] = sample_df["org_state_full"].apply(normalize_state_name)
sample_df["real_state_norm"] = sample_df["real_state"].apply(normalize_state_name)

# Find mismatches
mismatch = sample_df[
    (sample_df["real_state_norm"].notna()) &
    (sample_df["org_state_norm"].notna()) &
    (sample_df["real_state_norm"] != sample_df["org_state_norm"])
]

print(f"⚠️ Found {len(mismatch)} organizations with inconsistent states.\n")
display(mismatch[["org_name", "org_state_norm", "real_state_norm"]])
"""

'\n\n\ngeolocator = Nominatim(user_agent="nsf_org_checker")\n\ndef get_state_from_geopy(org_name):\n    try:\n        location = geolocator.geocode(org_name, addressdetails=True, timeout=10)\n        if location and "address" in location.raw:\n            address = location.raw["address"]\n            state = address.get("state") or address.get("state_code")\n            return state\n    except Exception:\n        pass\n    return None\n\nsample_df = data.copy() \n\nsample_df["real_state"] = sample_df["org_name"].apply(get_state_from_geopy)\n\ndef normalize_state_name(state):\n    if not state:\n        return None\n    state_lookup = us.states.lookup(state)\n    return state_lookup.name if state_lookup else state\n\nsample_df["org_state_norm"] = sample_df["org_state_full"].apply(normalize_state_name)\nsample_df["real_state_norm"] = sample_df["real_state"].apply(normalize_state_name)\n\n# Find mismatches\nmismatch = sample_df[\n    (sample_df["real_state_norm"].notna()) &\n    (sample

Now, since the library is based on search it can make mistakes. I checked manually all these univeritse and they all checked out, the bot made a mistake. 
There is one weird organization SHAKOPEE MDEWAKANTON SIOUX COMMUNITY.

We will now add the population by state

In [17]:
POP_PATH = CLEAN_DATA_DIR / "us_state_population.csv"
us_state_population = pd.read_csv(POP_PATH)
print(us_state_population.columns.tolist)
data = data.merge(
    us_state_population[["State", "state_population", "political_status"]],
    left_on="org_state_full",
    right_on="State",
    how="left"
)
print(data.columns.tolist)


<bound method IndexOpsMixin.tolist of Index(['State', 'state_population', 'political_status'], dtype='object')>
<bound method IndexOpsMixin.tolist of Index(['grant_id', 'status', 'reinstated', 'project_title', 'abstract',
       'org_name', 'org_state', 'nsf_total_budget', 'org_state_full', 'State',
       'state_population', 'political_status'],
      dtype='object')>


In [18]:
display(data[["org_state_full", "state_population", "political_status"]].head())


org_state_full state_population political_status
0   Pennsylvania       12,961,683       Republican
1          Texas       30,503,301       Republican
2     New Mexico        2,114,371       Democratic
3     California       38,965,193       Democratic
4       Oklahoma        4,053,824       Republican

In [19]:
describe_dataset(data)

Column            Type  Nulls  Null %  Unique      Min  \
0           grant_id           Int64      0    0.00    1970  1231319   
1             status  string[python]      0    0.00       2     None   
2         reinstated         boolean      0    0.00       2    False   
3      project_title          object      0    0.00    1649     None   
4           abstract          object      0    0.00    1665     None   
5           org_name  string[python]      0    0.00     506     None   
6          org_state        category      0    0.00      52     None   
7   nsf_total_budget           int64      0    0.00    1642     6774   
8     org_state_full          object      0    0.00      52     None   
9              State          object      2    0.10      51     None   
10  state_population          object      2    0.10      51     None   
11  political_status          object     11    0.56       2     None   

         Max         Mean          Std  \
0    2531008  2243661.047   148987.307   
1       None          NaN          NaN   
2       True        0.213        0.410   
3       None          NaN          NaN   
4       None          NaN          NaN   
5       None          NaN          NaN   
6       None          NaN          NaN   
7   25000000   871612.779  1305502.115   
8       None          NaN          NaN   
9       None          NaN          NaN   
10      None          NaN          NaN   
11      None          NaN          NaN   

                                       Example values  
0                         [2201796, 2315095, 2236163]  
1                            [Terminated, Reinstated]  
2                                       [False, True]  
3   [collaborative research investigating gender d...  
4   [despite evidence that gender differences in m...  
5   [Carnegie-Mellon University, University of Tex...  
6                                        [PA, TX, NM]  
7                          [1051218, 3641116, 609429]  
8                   [Pennsylvania, Texas, New Mexico]  
9                   [Pennsylvania, Texas, New Mexico]  
10                [12,961,683, 30,503,301, 2,114,371]  
11                           [Republican, Democratic]

In [20]:
missing_pop = data[data["political_status"].isna()]

print(f"Organizations with missing state population: {len(missing_pop)}\n")
display(missing_pop[["org_name", "org_state", "State", "org_state_full", "political_status"]].head(20))

Organizations with missing state population: 11



org_name org_state  \
237               University of Puerto Rico-Rio Piedras        PR   
242                           University of Puerto Rico        PR   
587   University of Puerto Rico Medical Sciences Campus        PR   
1004                 University of Puerto Rico Mayaguez        PR   
1227                   University of The Virgin Islands        VI   
1387                   University of The Virgin Islands        VI   
1529           University of Puerto Rico-Arecibo Campus        PR   
1541              University of Puerto Rico-Rio Piedras        PR   
1657              UNIVERSITY OF PUERTO RICO AT CAROLINA        PR   
1760              University of Puerto Rico-Rio Piedras        PR   
1824              University of Puerto Rico-Rio Piedras        PR   

            State  org_state_full political_status  
237   Puerto Rico     Puerto Rico              NaN  
242   Puerto Rico     Puerto Rico              NaN  
587   Puerto Rico     Puerto Rico              NaN  
1004  Puerto Rico     Puerto Rico              NaN  
1227          NaN  Virgin Islands              NaN  
1387          NaN  Virgin Islands              NaN  
1529  Puerto Rico     Puerto Rico              NaN  
1541  Puerto Rico     Puerto Rico              NaN  
1657  Puerto Rico     Puerto Rico              NaN  
1760  Puerto Rico     Puerto Rico              NaN  
1824  Puerto Rico     Puerto Rico              NaN

This are NOT a USA state, it is a USA territory, therefore we will drop these 11 rows. We have also noticed that there is the D.C which is neither a state nor a territory, however for the sake of visualization we belive that since it is located in the USA peninsula we can keep it.

In [21]:
data = data.dropna()

In [22]:
describe_dataset(data)

Column            Type  Nulls  Null %  Unique      Min  \
0           grant_id           Int64      0     0.0    1959  1231319   
1             status  string[python]      0     0.0       2     None   
2         reinstated         boolean      0     0.0       2    False   
3      project_title          object      0     0.0    1642     None   
4           abstract          object      0     0.0    1658     None   
5           org_name  string[python]      0     0.0     499     None   
6          org_state        category      0     0.0      50     None   
7   nsf_total_budget           int64      0     0.0    1635     6774   
8     org_state_full          object      0     0.0      50     None   
9              State          object      0     0.0      50     None   
10  state_population          object      0     0.0      50     None   
11  political_status          object      0     0.0       2     None   

         Max         Mean          Std  \
0    2531008  2243762.447   148703.097   
1       None          NaN          NaN   
2       True        0.214        0.411   
3       None          NaN          NaN   
4       None          NaN          NaN   
5       None          NaN          NaN   
6       None          NaN          NaN   
7   25000000   868922.155  1303328.684   
8       None          NaN          NaN   
9       None          NaN          NaN   
10      None          NaN          NaN   
11      None          NaN          NaN   

                                       Example values  
0                         [2201796, 2315095, 2236163]  
1                            [Terminated, Reinstated]  
2                                       [False, True]  
3   [collaborative research investigating gender d...  
4   [despite evidence that gender differences in m...  
5   [Carnegie-Mellon University, University of Tex...  
6                                        [PA, TX, NM]  
7                          [1051218, 3641116, 609429]  
8                   [Pennsylvania, Texas, New Mexico]  
9                   [Pennsylvania, Texas, New Mexico]  
10                [12,961,683, 30,503,301, 2,114,371]  
11                           [Republican, Democratic]

In [23]:
print(list(data["org_state_full"].unique()))

['Pennsylvania', 'Texas', 'New Mexico', 'California', 'Oklahoma', 'Alabama', 'Nevada', 'Washington', 'Maryland', 'Minnesota', 'North Dakota', 'South Carolina', 'Nebraska', 'Illinois', 'Michigan', 'Utah', 'Kentucky', 'Arizona', 'New York', 'Idaho', 'Ohio', 'North Carolina', 'Virginia', 'Georgia', 'Indiana', 'Arkansas', 'Colorado', 'Missouri', 'District of Columbia', 'Oregon', 'New Hampshire', 'Massachusetts', 'Tennessee', 'Louisiana', 'New Jersey', 'Wisconsin', 'Maine', 'Florida', 'South Dakota', 'Mississippi', 'Kansas', 'Connecticut', 'Hawaii', 'Montana', 'Delaware', 'Rhode Island', 'Iowa', 'Alaska', 'West Virginia', 'Vermont']


We are missing the state of Wyoming. I don't know what can we do with this.

Now we will juse add the Cruz list in the dataset.

In [24]:
LIST_PATH = CLEAN_DATA_DIR / "cruz_list.csv"
cruz_list = pd.read_csv(LIST_PATH, sep=";")
print(cruz_list.columns.tolist())
data = data.merge(
    cruz_list,
    left_on="grant_id",
    right_on="grant_number",
    how="left"
)

['grant_number', 'in_cruz_list']


In [25]:
describe_dataset(data)

Column            Type  Nulls  Null %  Unique        Min  \
0           grant_id           Int64      0    0.00    1959    1231319   
1             status  string[python]      0    0.00       2       None   
2         reinstated         boolean      0    0.00       2      False   
3      project_title          object      0    0.00    1642       None   
4           abstract          object      0    0.00    1658       None   
5           org_name  string[python]      0    0.00     499       None   
6          org_state        category      0    0.00      50       None   
7   nsf_total_budget           int64      0    0.00    1635       6774   
8     org_state_full          object      0    0.00      50       None   
9              State          object      0    0.00      50       None   
10  state_population          object      0    0.00      50       None   
11  political_status          object      0    0.00       2       None   
12      grant_number         float64    923   47.12    1036  2011780.0   
13      in_cruz_list          object    923   47.12       2       None   

          Max         Mean          Std  \
0     2531008  2243762.447   148703.097   
1        None          NaN          NaN   
2        True        0.214        0.411   
3        None          NaN          NaN   
4        None          NaN          NaN   
5        None          NaN          NaN   
6        None          NaN          NaN   
7    25000000   868922.155  1303328.684   
8        None          NaN          NaN   
9        None          NaN          NaN   
10       None          NaN          NaN   
11       None          NaN          NaN   
12  2520318.0  2266413.347   114453.991   
13       None          NaN          NaN   

                                       Example values  
0                         [2201796, 2315095, 2236163]  
1                            [Terminated, Reinstated]  
2                                       [False, True]  
3   [collaborative research investigating gender d...  
4   [despite evidence that gender differences in m...  
5   [Carnegie-Mellon University, University of Tex...  
6                                        [PA, TX, NM]  
7                          [1051218, 3641116, 609429]  
8                   [Pennsylvania, Texas, New Mexico]  
9                   [Pennsylvania, Texas, New Mexico]  
10                [12,961,683, 30,503,301, 2,114,371]  
11                           [Republican, Democratic]  
12                  [2201796.0, 2315095.0, 2236163.0]  
13                                      [True, False]

In [26]:
describe_dataset(cruz_list)

Column   Type  Nulls  Null %  Unique      Min      Max         Mean  \
0  grant_number  int64      0     0.0    1041  2011780  2520318  2266765.392   
1  in_cruz_list   bool      0     0.0       2    False     True        0.451   

          Std               Example values  
0  114388.845  [2011780, 2027519, 2034824]  
1       0.498                [False, True]

In [35]:
true_count_data = data["in_cruz_list"].sum()
print(f"✅ Number of True values in 'data': {true_count_data}")
true_count_list = cruz_list["in_cruz_list"].sum()
print(f"✅ Number of True values in 'cruz_list': {true_count_list}")

print(f"📊 Total number of grants in 'data': {len(data)}")

if "status" in data.columns and "in_cruz_list" in data.columns:
    reinstated_true_count = data[
        (data["status"] == "Reinstated") & (data["in_cruz_list"] == True)
    ].shape[0]

    print(f"🔄 Number of grants that are Reinstated AND True in cruz: {reinstated_true_count}")
else:
    print("⚠️ Columns 'status' or 'in_cruz_list' not found in data.")


✅ Number of True values in 'data': 1389
✅ Number of True values in 'cruz_list': 469
📊 Total number of grants in 'data': 1959
🔄 Number of grants that are Reinstated AND True in cruz: 368


The grants that are not on the dataset they for sure are NOT in cruz list. Therefore we change all NULL -> False

In [28]:
data["in_cruz_list"] = data["in_cruz_list"].apply(lambda x: False if x==None or x==False else True)
data = data.drop(columns=["grant_number"])
describe_dataset(data)

Column            Type  Nulls  Null %  Unique      Min  \
0           grant_id           Int64      0     0.0    1959  1231319   
1             status  string[python]      0     0.0       2     None   
2         reinstated         boolean      0     0.0       2    False   
3      project_title          object      0     0.0    1642     None   
4           abstract          object      0     0.0    1658     None   
5           org_name  string[python]      0     0.0     499     None   
6          org_state        category      0     0.0      50     None   
7   nsf_total_budget           int64      0     0.0    1635     6774   
8     org_state_full          object      0     0.0      50     None   
9              State          object      0     0.0      50     None   
10  state_population          object      0     0.0      50     None   
11  political_status          object      0     0.0       2     None   
12      in_cruz_list            bool      0     0.0       2    False   

         Max         Mean          Std  \
0    2531008  2243762.447   148703.097   
1       None          NaN          NaN   
2       True        0.214        0.411   
3       None          NaN          NaN   
4       None          NaN          NaN   
5       None          NaN          NaN   
6       None          NaN          NaN   
7   25000000   868922.155  1303328.684   
8       None          NaN          NaN   
9       None          NaN          NaN   
10      None          NaN          NaN   
11      None          NaN          NaN   
12      True        0.709        0.454   

                                       Example values  
0                         [2201796, 2315095, 2236163]  
1                            [Terminated, Reinstated]  
2                                       [False, True]  
3   [collaborative research investigating gender d...  
4   [despite evidence that gender differences in m...  
5   [Carnegie-Mellon University, University of Tex...  
6                                        [PA, TX, NM]  
7                          [1051218, 3641116, 609429]  
8                   [Pennsylvania, Texas, New Mexico]  
9                   [Pennsylvania, Texas, New Mexico]  
10                [12,961,683, 30,503,301, 2,114,371]  
11                           [Republican, Democratic]  
12                                      [True, False]

In [29]:
data.to_csv(TARGET_PATH, index=False)
